File: 05/pierpont/Find Peak Element.java

public class number_four {
	public static void main(String[] args) {
		int[] s = {1, 2, 3, 1};
		int a = findPeakElement(s);
		System.out.println(a);
	}
	
    public static int findPeakElement(int[] nums) {
    	for (int i = 1;i < nums.length - 1;i++) {
    		 if ((nums[i - 1] < nums[i])&&(nums[i + 1] < nums[i])) {
    			 return i;
    		 }		
    	}
		return 0;
        
    }

}

File: 05/piggy/find_peak_element.c
#include <stdio.h>

int findPeakElement(int *, int);

int main(void) {
    int a[4] = {1, 2, 3, 1};

    int peak = findPeakElement(a, 4);

    printf("peak = %d\n", peak);

    return 0;
}

int findPeakElement(int* nums, int numsSize) {
    if (numsSize == 2)
        return nums[0] > nums[1] ? 0 : 1;

    int middle = (numsSize - 1) / 2;

    if (nums[middle - 1] > nums[middle]) {
        for (int i = middle - 2, j = middle - 1; j >= 0; i--, j--) {
            if (i < 0)
                return j;
            if (nums[i] < nums[j])
                return j;
        } 
    } else if (nums[middle + 1] > nums[middle]) {
        for (int i = middle + 2, j = middle + 1; j < numsSize; i++, j++) {
            if (i >= numsSize)
                return j;
            if (nums[i] < nums[j])
                return j;
        }
    }

    return middle;
}
